### DISCUSSION TITLE
[share & question]general operating procedures

### DISCUSSION THREAD

**Discussion by DuncanDai** [bingogome/samm#58] [General]
### Background
I've seen the [demo](https://www.youtube.com/watch?v=vZK45noZVIA) in the form of a video (if there is a notebook demo, it would be nice). I tried to figure out the inconsistencies in the observed phenomena during the operation.

### Env
Ubuntu: 18.04
PyTorch: CUDA 11.1 + pytorch 1.10.1 + torchvision 0.11.2 + python 3.9.12
GPU: RTX 3090

### Operation and observed phenomena 
1 In the `Initialization` tab, import the data and calculate the Embedding (waiting for the server to compute the embedding for all slices and save the output). Then switch to the `Prompts` tab.    
2 Click on `Start Mask Sync` (If it clicked prematurely during the iteration of data in the `Red` window in the first step, 3D Slicer crashes. If it clicked before the server completes the computation and selects a prompt, there is no response). Choose a prompt, and you can observe real-time segmentation based on the prompt.      
3 After being satisfied with the segmentation result of the current slice, click `Freeze`, then move to another slice, repeat the process of selecting a prompt, and observe real-time segmentation results. Adjust prompts and repeat the `Freeze` operation.  

### Questions
1 In the operation process, if a certain procedure is not followed and `Start Mask Sync` is clicked, 3D Slicer crashes. I need to further summarize and update the info on the reproducible process.    
2 When should `Stop Mask Sync`, `Freeze`, and `Unfreeze` be used? Furthermore, for these operations, including whether it is possible to provide a **status indication** to indicate the state at the moment.    
3 Why should the prompt(s) from the previous slice be "inherited" in the current slice? Note that sometimes the prompt from one slice may fall outside the boundary for other slices. How can it be set so that the prompt between each slice is independent? Is it possible to propagate the model by selecting only one slice and updating all masks, similar to [Eiseg](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.6/EISeg/README_EN.md), and then adjust the prompt individually for poor segmentation results?  

Thank you in advance for your patience and support. Your work is really meaningful.

**Comment by bingogome** [comment]
Great post! 

You have pretty much nailed the procedure so I will just comment on the questions. 

1. One error that happens frequently is #34. Others being either the procedure was not followed or specific to MacOS #28. Much of it may be mitigated by completing #10. Though, the procedure is definitely something that needs to be improved on. For example, your question 2:

2. Great idea! Unfreeze and freeze is to do exactly what you had in the notes, so a slice-status indicator is a good idea when a slice was frozen. 

3. The prompts propagation for now is this: you have dropped prompts on a slice, and when you scroll to another slice, the prompts move along the same axis. You could still move the prompts around in the next slice if needed. You could use it in combination with the unfreeze and freeze. My understanding the use case with SAM in medical images is that it's a good semi-automatic labeling tool. Full propagation is something I need to think more on.

**Comment by bingogome** [comment]
The error with crashes and view mismatch are addressed in the most recent changes.
